src/main/java/co/cask/gcp/bigquery/BigQuerySink.java

```diff
@@ -180,116 +114,121 @@ public void prepareRun(BatchSinkContext context) throws Exception {
       BigQueryFileFormat.NEWLINE_DELIMITED_JSON,
       TextOutputFormat.class);
 
-    LineageRecorder lineageRecorder = new LineageRecorder(context, config.referenceName);
-    lineageRecorder.createExternalDataset(config.getSchema());
+    setOutputFormat(context);
+    emitLineage(context, fields);
+  }
 
-    context.addOutput(Output.of(config.referenceName, new OutputFormatProvider() {
-      @Override
-      public String getOutputFormatClassName() {
-        return IndirectBigQueryOutputFormat.class.getName();
-      }
+  @Override
+  public void initialize(BatchRuntimeContext context) throws Exception {
+    super.initialize(context);
+    schema = config.getSchema();
+  }
 
-      @Override
-      public Map<String, String> getOutputFormatConfiguration() {
-        Map<String, String> config = new HashMap<>();
-        for (Map.Entry<String, String> entry : configuration) {
-          config.put(entry.getKey(), entry.getValue());
+  @Override
+  public void transform(StructuredRecord input, Emitter<KeyValue<JsonObject, NullWritable>> emitter) throws Exception {
+    List<Schema.Field> fields = config.getSchema().getFields();
+    JsonObject object = new JsonObject();
+    for (Schema.Field field : fields) {
+      BigQueryUtils.validateSimpleType(field);
+      decodeSimpleTypes(object, field.getName(), input);
+    }
+    emitter.emit(new KeyValue<>(object, NullWritable.get()));
+  }
+
+  @Override
+  public void onRunFinish(boolean succeeded, BatchSinkContext context) {
+    if (config.bucket == null) {
+      Path gcsPath = new Path(String.format("gs://%s", uuid.toString()));
+      try {
+        FileSystem fs = gcsPath.getFileSystem(configuration);
+        if (fs.exists(gcsPath)) {
+          fs.delete(gcsPath, true);
         }
-        return config;
+      } catch (IOException e) {
+        LOG.warn("Failed to delete bucket " + gcsPath.toUri().getPath() + ", " + e.getMessage());
       }
-    }));
-
-    if (!fields.isEmpty()) {
-      lineageRecorder.recordWrite("Write", "Wrote to BigQuery table.",
-                                  fields.stream().map(BigQueryTableFieldSchema::getName).collect(Collectors.toList()));
     }
   }
 
-  @Override
-  public void initialize(BatchRuntimeContext context) throws Exception {
-    super.initialize(context);
-    schema = config.getSchema();
+  private String getBucket() {
```

Comment:
most people don't expect a getter to modify state. It would be better to name this 'setBucket' and pass in the Configuration object.

I also don't think that Configuration should be a class variable. It should only live in the prepareRun() method, otherwise it's confusing when it is null and when it is non-null.